indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java

```diff
@@ -118,9 +117,24 @@ public void stop()
 
   public Optional<SupervisorReport> getSupervisorStatus(String id)
   {
-    return supervisors.get(id) == null
-           ? Optional.<SupervisorReport>absent()
-           : Optional.fromNullable(supervisors.get(id).lhs.getStatus());
+    Pair<Supervisor, SupervisorSpec> supervisor = supervisors.get(id);
+    return supervisor == null ? Optional.<SupervisorReport>absent() : Optional.fromNullable(supervisor.lhs.getStatus());
+  }
+
+  private boolean possiblyStopAndRemoveSupervisorInternal(String id, boolean writeTombstone)
+  {
+    Pair<Supervisor, SupervisorSpec> pair = supervisors.get(id);
+    if (pair != null) {
+      if (writeTombstone) {
+        metadataSupervisorManager.insert(id, new NoopSupervisorSpec()); // where NoopSupervisorSpec is a tombstone
+      }
+      pair.lhs.stop(true);
+      supervisors.remove(id);
```

Comment:
@drcrallen hm, my thought in ordering it this way is that if I removed it from the list first and then tried to stop it, if stop() threw an exception that prevented the supervisor from actually stopping then we would have already removed it from our list and could no longer message it (or we could create a 2nd supervisor that would conflict with the zombie one). With this ordering, it will first remove it from metadata storage (so that even if stop does fail it won't restart on next run), then it'll try to stop it, and only if stop succeeds will it remove it from the list.

For the Kafka implementation, this is mostly moot since stop() shouldn't throw an exception, but since SupervisorManager can manage other types of supervisors as well, it seemed better to make sure it actually stopped before removing it (in case some supervisors could have transient issues that would prevent them from stopping but calling it a 2nd time might succeed?)